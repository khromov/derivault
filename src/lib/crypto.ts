import { get } from 'svelte/store';
import { computationIntensity } from './stores';
import { mnemonicToSeed } from 'web-bip39';

const ITERATIONS_PER_INTENSITY = 250000; // 250.000 iterations per intensity

export async function deriveMasterKey(
	passphrase: string,
	intensityOverride?: number
): Promise<Uint8Array> {
	const encoder = new TextEncoder();
	const passphraseBuffer = encoder.encode(passphrase);
	const salt = encoder.encode('ConstantSaltForDeterministicResults');

	const keyMaterial = await window.crypto.subtle.importKey(
		'raw',
		passphraseBuffer,
		{ name: 'PBKDF2' },
		false,
		['deriveBits']
	);

	const intensity = intensityOverride ?? get(computationIntensity);
	const derivedBits = await window.crypto.subtle.deriveBits(
		{
			name: 'PBKDF2',
			salt: salt,
			iterations: ITERATIONS_PER_INTENSITY * intensity,
			hash: 'SHA-512'
		},
		keyMaterial,
		256
	);

	return new Uint8Array(derivedBits);
}

export async function deriveBip39MasterKey(
	mnemonic: string,
	intensityOverride?: number
): Promise<Uint8Array> {
	// First get the BIP39 seed
	const bip39Seed = await mnemonicToSeed(mnemonic.trim());

	// Import the seed for additional key stretching
	const keyMaterial = await window.crypto.subtle.importKey(
		'raw',
		bip39Seed,
		{ name: 'PBKDF2' },
		false,
		['deriveBits']
	);

	// Add our own stretching with configurable intensity
	const intensity = intensityOverride ?? get(computationIntensity);
	const extraStretchedKey = await window.crypto.subtle.deriveBits(
		{
			name: 'PBKDF2',
			salt: new TextEncoder().encode('DeriVault-BIP39-Extra-Stretching'),
			iterations: ITERATIONS_PER_INTENSITY * intensity,
			hash: 'SHA-512'
		},
		keyMaterial,
		256
	);

	return new Uint8Array(extraStretchedKey);
}

export async function generatePassword(
	masterKey: Uint8Array,
	site: { email: string; domain: string; rotationRounds: number }
): Promise<string> {
	const encoder = new TextEncoder();
	const siteData = encoder.encode(`${site.email}:${site.domain}:${site.rotationRounds}`);

	// First, derive a site-specific key using PBKDF2 with the master key as the base
	const siteKeyMaterial = await window.crypto.subtle.importKey(
		'raw',
		masterKey,
		{ name: 'PBKDF2' },
		false,
		['deriveBits']
	);

	const intensity = get(computationIntensity);
	const siteKey = await window.crypto.subtle.deriveBits(
		{
			name: 'PBKDF2',
			salt: siteData, // Use site-specific data as salt
			iterations: ITERATIONS_PER_INTENSITY * intensity, // Make it as slow as the master key derivation
			hash: 'SHA-512'
		},
		siteKeyMaterial,
		512 // Use larger output for more entropy
	);

	// Convert the derived key into a password
	const hashArray = Array.from(new Uint8Array(siteKey));
	const hashHex = hashArray.map((b) => b.toString(16).padStart(2, '0')).join('');

	const allChars =
		'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()_+-=[]{}|;:,.<>?';

	let password = '';
	for (let i = 0; i < 16; i++) {
		const charIndex = parseInt(hashHex.substr(i * 2, 2), 16) % allChars.length;
		password += allChars[charIndex];
	}

	return password;
}
