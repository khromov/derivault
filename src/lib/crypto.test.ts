import { describe, it, expect } from 'vitest';
import { deriveMasterKey, generatePassword } from './crypto';

describe('deriveMasterKey', () => {
	it('produces the same key for the same passphrase', async () => {
		const passphrase = 'test-passphrase';
		const intensityOverride = 1; // Use low intensity for faster tests

		const key1 = await deriveMasterKey(passphrase, intensityOverride);
		const key2 = await deriveMasterKey(passphrase, intensityOverride);

		expect(Array.from(key1)).toEqual(Array.from(key2));
	});

	it('produces different keys for different passphrases', async () => {
		const passphrase1 = 'test-passphrase-1';
		const passphrase2 = 'test-passphrase-2';
		const intensityOverride = 1;

		const key1 = await deriveMasterKey(passphrase1, intensityOverride);
		const key2 = await deriveMasterKey(passphrase2, intensityOverride);

		expect(Array.from(key1)).not.toEqual(Array.from(key2));
	});

	it('produces different keys for different intensities', async () => {
		const passphrase = 'test-passphrase';

		const key1 = await deriveMasterKey(passphrase, 1);
		const key2 = await deriveMasterKey(passphrase, 2);

		expect(Array.from(key1)).not.toEqual(Array.from(key2));
	});
});

describe('generatePassword', () => {
	it('generates the same password for same inputs', async () => {
		const masterKey = new Uint8Array([1, 2, 3, 4, 5]); // Simplified key for testing
		const site = {
			email: 'test@example.com',
			domain: 'example.com',
			rotationRounds: 1
		};

		const password1 = await generatePassword(masterKey, site);
		const password2 = await generatePassword(masterKey, site);

		expect(password1).toEqual(password2);
	});

	it('generates different passwords for different emails', async () => {
		const masterKey = new Uint8Array([1, 2, 3, 4, 5]);
		const site1 = {
			email: 'test1@example.com',
			domain: 'example.com',
			rotationRounds: 1
		};
		const site2 = {
			email: 'test2@example.com',
			domain: 'example.com',
			rotationRounds: 1
		};

		const password1 = await generatePassword(masterKey, site1);
		const password2 = await generatePassword(masterKey, site2);

		expect(password1).not.toEqual(password2);
	});

	it('generates different passwords for different domains', async () => {
		const masterKey = new Uint8Array([1, 2, 3, 4, 5]);
		const site1 = {
			email: 'test@example.com',
			domain: 'example1.com',
			rotationRounds: 1
		};
		const site2 = {
			email: 'test@example.com',
			domain: 'example2.com',
			rotationRounds: 1
		};

		const password1 = await generatePassword(masterKey, site1);
		const password2 = await generatePassword(masterKey, site2);

		expect(password1).not.toEqual(password2);
	});

	it('generates different passwords for different rotation rounds', async () => {
		const masterKey = new Uint8Array([1, 2, 3, 4, 5]);
		const site1 = {
			email: 'test@example.com',
			domain: 'example.com',
			rotationRounds: 1
		};
		const site2 = {
			email: 'test@example.com',
			domain: 'example.com',
			rotationRounds: 2
		};

		const password1 = await generatePassword(masterKey, site1);
		const password2 = await generatePassword(masterKey, site2);

		expect(password1).not.toEqual(password2);
	});

	it('generates passwords with proper length and character types', async () => {
		const masterKey = new Uint8Array([1, 2, 3, 4, 5]);
		const site = {
			email: 'test@example.com',
			domain: 'example.com',
			rotationRounds: 1
		};

		const password = await generatePassword(masterKey, site);

		expect(password).toMatch(/^[a-zA-Z0-9!@#$%^&*()_+\-=[\]{}|;:,.<>?]{16}$/);
		expect(password.length).toBe(16);
	});

	describe('snapshot tests', () => {
		// Helper function to create consistent master keys for snapshot tests
		const createMasterKey = (seed: number): Uint8Array => {
			return new Uint8Array(Array.from({ length: 32 }, (_, i) => (seed + i) % 256));
		};

		it('generates expected password for default case', async () => {
			const masterKey = createMasterKey(1);
			const site = {
				email: 'user@example.com',
				domain: 'example.com',
				rotationRounds: 1
			};
			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"|J@OE)MPLIW**QCk"`);
		});

		it('generates expected password for Gmail account', async () => {
			const masterKey = createMasterKey(2);
			const site = {
				email: 'user@gmail.com',
				domain: 'google.com',
				rotationRounds: 1
			};

			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"^G%!-,4f[l0Q!?;0"`);
		});

		it('generates expected password with high rotation round', async () => {
			const masterKey = createMasterKey(3);
			const site = {
				email: 'admin@company.com',
				domain: 'company.com',
				rotationRounds: 10
			};

			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"d+r[4AHm;[7C8X6E"`);
		});

		it('generates expected password with special characters in email', async () => {
			const masterKey = createMasterKey(4);
			const site = {
				email: 'user+tag@example.com',
				domain: 'example.com',
				rotationRounds: 1
			};

			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"8DV9ox!Z&qe{ARS%"`);
		});

		it('generates expected password with subdomain', async () => {
			const masterKey = createMasterKey(5);
			const site = {
				email: 'user@example.com',
				domain: 'sub.example.com',
				rotationRounds: 1
			};

			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"W]$Fv|3)U+|1|ojO"`);
		});

		it('generates expected password with very long inputs', async () => {
			const masterKey = createMasterKey(6);
			const site = {
				email: 'very.long.email.address.with.many.parts@really.long.domain.example.com',
				domain: 'extremely.long.subdomain.with.many.parts.example.com',
				rotationRounds: 5
			};

			const password = await generatePassword(masterKey, site);
			expect(password).toMatchInlineSnapshot(`"!znIhwZjqqP>Gg!C"`);
		});
	});
});
