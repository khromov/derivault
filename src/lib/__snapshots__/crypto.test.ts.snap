// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deriveBip39MasterKey > generates expected key for standard test vector 1`] = `"fa1d8c30eacbaa40b142ffcfef8fbf1e68e3e8ceeefe64da2eb1e8a9ffac0bbd"`;

exports[`generatePassword > snapshot tests > generates expected password for Gmail account 1`] = `"r)Bbra:jtVW}.zi-"`;

exports[`generatePassword > snapshot tests > generates expected password for default case 1`] = `")7+-bRND*G;_&k0e"`;

exports[`generatePassword > snapshot tests > generates expected password with high rotation round 1`] = `"ID2b)_?Eu$%+@OEe"`;

exports[`generatePassword > snapshot tests > generates expected password with special characters in email 1`] = `"$2zd}F5y^S%Mvv6w"`;

exports[`generatePassword > snapshot tests > generates expected password with subdomain 1`] = `"Lt0W%Xe.C?7m6TeD"`;

exports[`generatePassword > snapshot tests > generates expected password with very long inputs 1`] = `"AB#8^#)roL:b4vSU"`;
